### PR TITLE
feat: Implement Default Values for UuidValue field types in Models

### PR DESCRIPTION
**Pull request by klkucaj**
### Overview

This is a follow-up to #2383.  
This PR introduces support for default values in the `UuidValue` field type.

### Default Types

There are three types of default values that can be specified:

- **defaultDatabase**: Adds a default value to the database table column.
- **defaultModel**: Adds a default value to the model field.
- **default**: Adds default values to both the model field and the database table column.

### Default Value Options for `UuidValue` Fields

Default values can be assigned to `UuidValue` fields as follows:

- **UUID String Values**:
  - **Dart Side**: A valid UUID string can be specified, for example, `'550e8400-e29b-41d4-a716-446655440000'`.
  - **Database Side**: The corresponding UUID string value will be stored in the database column, e.g., `'550e8400-e29b-41d4-a716-446655440000'`.

- **Random UUID**:
  - **Dart Side**: To generate a random UUID, use the keyword `"random"`. This will be internally handled by `Uuid().v4obj()` to generate a UUID at runtime.
  - **Database Side**: The random UUID value will be stored in the database column using the `gen_random_uuid()` function in PostgreSQL.

### Related Issues

- #1140
- #236

## Pre-launch Checklist

- [x] I read the [Contribute](https://docs.serverpod.dev/contribute) page and followed the process outlined there for submitting PRs.
- [x] This update contains only one single feature or bug fix and nothing else. (If you are submitting multiple fixes, please make multiple PRs.)
- [x] I read and followed the [Dart Style Guide](https://dart.dev/guides/language/effective-dart/style) and formatted the code with [dart format](https://dart.dev/tools/dart-format).
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`), and made sure that the documentation follows the same style as other Serverpod documentation. I checked spelling and grammar.
- [x] I added new tests to check the change I am making.
- [x] All existing and new tests are passing.
- [x] Any breaking changes are documented below.

If you need help, consider asking for advice on the [discussion board](https://github.com/serverpod/serverpod/discussions).

## Breaking changes

### CODE REVIEW

File: tools/serverpod_cli/test/analyzer/models/stateful_analyzer/model_validation/field/default/uuid/field_uuid_default_model_test.dart [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+import 'package:serverpod_cli/src/test_util/builders/generator_config_builder.dart';
+import 'package:serverpod_cli/src/test_util/builders/model_source_builder.dart';
+import 'package:test/test.dart';
+
+void main() {
+  var config = GeneratorConfigBuilder().build();
+
+  group('Given a class with fields with a "defaultModel" keyword', () {
+    test(
+      'when the field is of type UUID and the defaultModel is set to "random", then the field should have a "default model" value',
```
[the remark this answers is not in the document]
Comment (SandPod) [reply tools/serverpod_cli/test/analyzer/models/stateful_analyzer/model_validation/field/default/uuid/field_uuid_default_model_test.dart:14]: Check the rest of the tests for this as well.

File: tools/serverpod_cli/test/analyzer/models/stateful_analyzer/model_validation/field/default/uuid/field_uuid_default_model_test.dart [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+        var collector = CodeGenerationCollector();
+        StatefulAnalyzer(config, models, onErrorsCollector(collector))
+            .validateAll();
+
+        expect(collector.errors, isNotEmpty);
+
+        var firstError = collector.errors.first as SourceSpanSeverityException;
+        expect(
+          firstError.message,
+          'The "defaultModel" value must be a valid UUID string or "random" (e.g., "defaultModel"=random or "defaultModel"="550e8400-e29b-41d4-a716-446655440000").',
```
Comment (SandPod) [inline tools/serverpod_cli/test/analyzer/models/stateful_analyzer/model_validation/field/default/uuid/field_uuid_default_model_test.dart:122]: Suggestion: Flip the order of the examples so that it matches the order of the hint.

Comment (SandPod) [reply tools/serverpod_cli/test/analyzer/models/stateful_analyzer/model_validation/field/default/uuid/field_uuid_default_model_test.dart:122]: Same for the other hints.

File: tools/serverpod_cli/test/analyzer/models/stateful_analyzer/model_validation/field/default/uuid/field_uuid_default_model_test.dart [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+    test(
+      'when the field is of type UUID with non-UUID defaultModel value, then an error is generated',
+      () {
+        var models = [
+          ModelSourceBuilder().withYaml(
+            '''
+          class: Example
+          table: example
+          fields:
+            uuidType: UuidValue?, defaultModel=test
```
Comment (SandPod) [inline tools/serverpod_cli/test/analyzer/models/stateful_analyzer/model_validation/field/default/uuid/field_uuid_default_model_test.dart:220]: Question: Is there a difference between this test and the one where `INVALID_UUID` is specified?

File: tools/serverpod_cli/test/analyzer/models/stateful_analyzer/model_validation/field/default/uuid/field_uuid_default_model_test.dart [hunk 4]
```diff
@@ -0,0 +1,267 @@
+import 'package:serverpod_cli/analyzer.dart';
+import 'package:serverpod_cli/src/analyzer/code_analysis_collector.dart';
+import 'package:serverpod_cli/src/analyzer/models/stateful_analyzer.dart';
+import 'package:serverpod_cli/src/generator/code_generation_collector.dart';
+import 'package:serverpod_cli/src/test_util/builders/generator_config_builder.dart';
+import 'package:serverpod_cli/src/test_util/builders/model_source_builder.dart';
+import 'package:test/test.dart';
+
+void main() {
```
Comment (SandPod) [inline tools/serverpod_cli/test/analyzer/models/stateful_analyzer/model_validation/field/default/uuid/field_uuid_default_model_test.dart:9]: Please check all comments in this file with the other analyzer tests files.

File: tools/serverpod_cli/lib/src/generator/types.dart [hunk 5]
```diff
@@ -454,6 +454,8 @@ class TypeDefinition {
         return DefaultValueAllowedType.double;
       case ValueType.string:
         return DefaultValueAllowedType.string;
+      case ValueType.uuidValue:
```
Comment (SandPod) [inline tools/serverpod_cli/lib/src/generator/types.dart:457]: Comment: This is part of the second commit but would have been required to be able to run the tests in the first commit.

File: tools/serverpod_cli/lib/src/database/create_definition.dart [hunk 6]
```diff
@@ -142,16 +142,17 @@ String? _getColumnDefault(
...
-      var defaultValue = column.defaultPersistValue;
       return '$defaultValue';
     case DefaultValueAllowedType.string:
-      var defaultValue = column.defaultPersistValue;
       return '$defaultValue::text';
+    case DefaultValueAllowedType.uuidValue:
+      if (defaultUuidValueRandom == defaultValue) {
+        return 'gen_random_uuid()';
+      }
+      return '$defaultValue::uuid';
```
Comment (SandPod) [inline tools/serverpod_cli/lib/src/database/create_definition.dart:156]: Question: Changes here look a bit funky, why could all `var defaultValue = column.defaultPersistValue;` be removed?

File: tests/serverpod_test_client/lib/src/protocol/defaults/uuid/uuid_default.dart [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+
+// ignore_for_file: no_leading_underscores_for_library_prefixes
+import 'package:serverpod_client/serverpod_client.dart' as _i1;
+import 'package:uuid/uuid.dart' as _i2;
+
+abstract class UuidDefault implements _i1.SerializableModel {
+  UuidDefault._({
+    this.id,
+    _i1.UuidValue? uuidDefaultRandom,
+    _i1.UuidValue? uuidDefaultRandomNull,
```
Comment (SandPod) [inline tests/serverpod_test_client/lib/src/protocol/defaults/uuid/uuid_default.dart:18]: Question(For my knowledge): Why are both of these nullable in the model? As far as I understand, only one of them should be nullable.

File: tests/serverpod_test_server/test/defaults/uuid/uuid_default_test.dart [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+  group(
+    "Given a class with 'default' UUID fields",
+    () {
+      test(
+        'when an object of the class is created, then the "default=random" UUID field should not be null',
+        () {
+          var object = UuidDefault();
+          expect(object.uuidDefaultRandom, isNotNull);
+        },
+      );
```
Comment (SandPod) [inline tests/serverpod_test_server/test/defaults/uuid/uuid_default_test.dart:15]: Suggestion: Add tests for the field `uuidDefaultRandomNull`.

This file does not contain any test for uuidDefaultRandomNull. Is that intentional?

Comment (SandPod) [reply tests/serverpod_test_server/test/defaults/uuid/uuid_default_test.dart:15]: This is the same for all tests in this directory.

**Review by SandPod — COMMENTED** [review]
Awesome work! 🙌 
Almost done with default values.

A couple of questions for you.

File: tools/serverpod_cli/lib/src/database/create_definition.dart (hunk 6 above)
Comment (klkucaj) [reply tools/serverpod_cli/lib/src/database/create_definition.dart:156]: Because we already have `var defaultValue = column.defaultPersistValue;` on line 129
Screenshot 2024-08-15 at 9 11 07 AM

File: tests/serverpod_test_client/lib/src/protocol/defaults/uuid/uuid_default.dart (hunk 7 above)
Comment (klkucaj) [reply tests/serverpod_test_client/lib/src/protocol/defaults/uuid/uuid_default.dart:18]: No, both should be nullable (optional) because they will have the default value on constructor initialization
Screenshot 2024-08-15 at 9 13 33 AM

but if you check the model field, only one of these is nullable:
Screenshot 2024-08-15 at 9 14 16 AM

File: tools/serverpod_cli/test/analyzer/models/stateful_analyzer/model_validation/field/default/uuid/field_uuid_default_model_test.dart (hunk 3 above)
Comment (klkucaj) [reply tools/serverpod_cli/test/analyzer/models/stateful_analyzer/model_validation/field/default/uuid/field_uuid_default_model_test.dart:220]: No, just adding extra test, should this or the other one be removed ?

File: tools/serverpod_cli/lib/src/generator/types.dart (hunk 5 above)
Comment (klkucaj) [reply tools/serverpod_cli/lib/src/generator/types.dart:457]: Yes, you are right, i missed this when i was working on the commit history, sorry!

File: tests/serverpod_test_server/test/defaults/uuid/uuid_default_test.dart (hunk 8 above)
Comment (klkucaj) [reply tests/serverpod_test_server/test/defaults/uuid/uuid_default_test.dart:15]: Nope is not intentional, i just missed this field, thank you!

File: tools/serverpod_cli/test/analyzer/models/stateful_analyzer/model_validation/field/default/uuid/field_uuid_default_model_test.dart (hunk 3 above)
Comment (klkucaj) [reply tools/serverpod_cli/test/analyzer/models/stateful_analyzer/model_validation/field/default/uuid/field_uuid_default_model_test.dart:220]: Done!

File: tools/serverpod_cli/lib/src/analyzer/models/utils/quote_utils.dart [hunk 9]
```diff
@@ -0,0 +1,43 @@
+/// Regular expression pattern that matches a string entirely enclosed
```
Comment (SandPod) [inline tools/serverpod_cli/lib/src/analyzer/models/utils/quote_utils.dart:1]: Suggestion: Add examples to illustrate what each method captures.

Awesome to have all of these in a single class.
But it hard to understand what would be matched by both the regex patterns but also what would return true for the methods.
I think some examples on each case would help future developers here.

File: tools/serverpod_cli/lib/src/analyzer/models/validation/restrictions/default.dart [hunk 10]
```diff
@@ -208,4 +212,94 @@ class DefaultValueRestriction extends ValueRestriction {
...
+          span,
+        ),
+      );
+      return errors;
+    }
+
+    /// Extract the actual UUID string by removing quotes
+    String uuidString = value.substring(1, value.length - 1);
+
+    bool isValidUuid = RegExp(
```
Comment (SandPod) [inline tools/serverpod_cli/lib/src/analyzer/models/validation/restrictions/default.dart:290]: Suggestion: Use UUID v4 validation here instead of your defined regexp.

It should be possible to just parse it here.

File: tools/serverpod_cli/test/analyzer/models/stateful_analyzer/model_validation/field/default/string/field_string_default_model_test.dart [hunk 11]
```diff
@@ -197,7 +196,7 @@ void main() {
         var definition = definitions.first as ClassDefinition;
         expect(
           definition.fields.last.defaultModelValue,
-          '"This \'is\' a default model value"',
+          '\'This \\\'is\\\' a default model value\'',
```
Comment (SandPod) [inline tools/serverpod_cli/test/analyzer/models/stateful_analyzer/model_validation/field/default/string/field_string_default_model_test.dart:199]: Question: Is this format of the string correctly parsed in Postgres?

Comment (SandPod) [reply tools/serverpod_cli/test/analyzer/models/stateful_analyzer/model_validation/field/default/string/field_string_default_model_test.dart:199]: If the value is defined in the database, would we get `This 'is' a default model value` as string for the object?

File: tools/serverpod_cli/test/analyzer/models/stateful_analyzer/model_validation/field/default/string/field_string_default_test.dart [hunk 12]
```diff
@@ -11,15 +11,15 @@ void main() {
 
   group('Given a class with fields with a "default" keyword', () {
```
Comment (SandPod) [inline tools/serverpod_cli/test/analyzer/models/stateful_analyzer/model_validation/field/default/string/field_string_default_test.dart:12]: chore: This files seems to have been changed into a UUID test file but it is actually a String test file. Please revert the changes here..

**Review by SandPod — CHANGES_REQUESTED** [review]
Couple of additional comments.

I would also like to see models defined that uses double quotes to ensure we actually support it.

I think it would be good to add some integration tests that validate that the escaped strings actually produced the desired output if the default values are set in the database.

File: tools/serverpod_cli/test/analyzer/models/stateful_analyzer/model_validation/field/default/string/field_string_default_model_test.dart (hunk 11 above)
Comment (klkucaj) [reply tools/serverpod_cli/test/analyzer/models/stateful_analyzer/model_validation/field/default/string/field_string_default_model_test.dart:199]: Done!

File: tests/serverpod_test_server/test/defaults/string/string_default_persist_test.dart [hunk 13]
```diff
@@ -21,6 +21,94 @@ void main() async {
           expect(object.stringDefaultPersist, equals('A specific value'));
         },
       );
+
```
Comment (SandPod) [inline tests/serverpod_test_server/test/defaults/string/string_default_persist_test.dart:24]: Question: What are we validating with these additional tests that are not validated with the first ones?

File: tests/serverpod_test_server/test_integration/database_operations/default/string/string_default_persist_test.dart [hunk 14]
```diff
@@ -55,5 +55,407 @@ void main() async {
...
+      () async {
+        var object = StringDefaultPersist();
+        var databaseObject = await StringDefaultPersist.db.insertRow(
+          session,
+          object,
+        );
+        expect(
+            databaseObject
+                .stringDefaultPersistSingleQuoteWithOneSingleEscapeQuote,
+            equals('This is a \'default persist value'));
```
Comment (SandPod) [inline tests/serverpod_test_server/test_integration/database_operations/default/string/string_default_persist_test.dart:72]: Suggestion(NIT): Add a comma at the end of the expect for better formatting.

Please apply this to all tests that have bad line breaking in this file.

**Review by SandPod — COMMENTED** [review]
Great work!

A couple of last NIT suggestions, comments and questions!
Ping me when you have answered and we should be good to go.

File: tools/serverpod_cli/lib/src/generator/dart/library_generators/model_library_generator.dart [hunk 15]
```diff
@@ -1050,6 +1050,18 @@ class SerializableModelLibraryGenerator {
         return literalNum(double.parse(defaultValue)).code;
       case DefaultValueAllowedType.string:
         return Code(defaultValue);
+      case DefaultValueAllowedType.uuidValue:
+        if (defaultValue is! String) return null;
```
[the remark this answers is not in the document]
Comment (klkucaj) [reply tools/serverpod_cli/lib/src/generator/dart/library_generators/model_library_generator.dart:1054]: Yes, done!

**Review by SandPod — APPROVED** [review]
LGTM! 🚀
